src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java

```diff
@@ -113,63 +117,134 @@
      * Creates a FeatureDataSource backed by the provided File that uses the provided codec to decode records
      * from that file, and assigns this data source a logical name. We will look ahead the specified number of bases
      * during queries that produce cache misses.
      *
-     * @param featureFile file containing Features
-     * @param codec codec with which to decode the records from featureFile
-     * @param name logical name for this data source (may be null)
+     * @param featureInput a feature input specifying a source of Features
      * @param queryLookaheadBases look ahead this many bases during queries that produce cache misses
+     * @param featureType when searching for {@link FeatureCodec} for this data source, restrict the search to codecs
+     *                              that produce this type of Feature
      */
-    public FeatureDataSource( final File featureFile, final FeatureCodec<T, ?> codec, final String name, final int queryLookaheadBases ) {
-        this.featureFile = Utils.nonNull(featureFile, "FeatureDataSource cannot be created from null feature file");
-        this.codec = Utils.nonNull(codec, "FeatureDataSource cannot be created from null codec");
+     FeatureDataSource(final FeatureInput<T> featureInput, final int queryLookaheadBases, final Class<? extends Feature> featureType) {
         Utils.validateArg( queryLookaheadBases >= 0, "Query lookahead bases must be >= 0");
+        this.featureInput = Utils.nonNull(featureInput, "featureInput must not be null");
+
+        // Create a feature reader without requiring an index.  We will require one ourselves as soon as
+        // a query by interval is attempted.
+        this.featureReader = getFeatureReader(featureInput, featureType);
+
+        if(isGenomicsDBPath(featureInput.getFeaturePath())) {
+            //genomics db uri's have no associated index file to read from, but they do support random access
+            this.hasIndex = false;
+            this.supportsRandomAccess = true;
+        } else if (featureReader instanceof AbstractFeatureReader) {
+            this.hasIndex = ((AbstractFeatureReader<T, ?>) featureReader).hasIndex();
+            this.supportsRandomAccess = hasIndex;
+        } else {
+            throw new GATKException("Found a feature input that was neither GenomicsDB or a Tribble AbstractFeatureReader.  Input was " + featureInput.toString() + ".");
+        }
+
 
-        if ( ! featureFile.canRead() || featureFile.isDirectory() ) {
-            throw new UserException.CouldNotReadInputFile("File " + featureFile.getAbsolutePath() + " does not exist, is unreadable, or is a directory");
+        this.currentIterator = null;
+        this.intervalsForTraversal = null;
+        this.queryCache = new FeatureCache<>();
+        this.queryLookaheadBases = queryLookaheadBases;}
+
+    public static boolean isGenomicsDBPath(final String filePath){
+        return Utils.nonNull(filePath).startsWith(GENOMIC_DB_URI_SCHEME);
```

Comment:
Might want this method to return false on `filePath == null` rather than than blowing up.